File: python/sorting/counting_sort.py
"""
Counting Sort
-------------
Time:  O(n + k) where k is the range of input values
Space: O(k)

Non-comparison based sort. Works great when k isn't much larger than n.
Only works with non-negative integers (or you need to shift the values).
"""


def counting_sort(arr):
    if not arr:
        return arr

    max_val = max(arr)
    count = [0] * (max_val + 1)

    for num in arr:
        count[num] += 1

    # rebuild the array from counts
    idx = 0
    for val, freq in enumerate(count):
        for _ in range(freq):
            arr[idx] = val
            idx += 1

    return arr

File: javascript/sorting/countingSort.js
/**
 * Counting Sort
 * Time:  O(n + k) where k = range of values
 * Space: O(k)
 *
 * Non-comparison sort. Only works with non-negative integers.
 */

function countingSort(arr) {
  if (arr.length === 0) return arr;

  const max = Math.max(...arr);
  const count = new Array(max + 1).fill(0);

  for (const num of arr) {
    count[num]++;
  }

  let idx = 0;
  for (let val = 0; val < count.length; val++) {
    for (let c = 0; c < count[val]; c++) {
      arr[idx] = val;
      idx++;
    }
  }

  return arr;
}

module.exports = countingSort;
